File: src/types/events.ts
/**
 * Internal Tauri Events for zblade
 * 
 * This module defines the event contract between zblade's Rust backend and React frontend.
 * Events flow in one direction: Backend → Frontend (via Tauri's emit)
 * Frontend → Backend communication uses invoke() commands instead.
 * 
 * These types must match the Rust definitions in src-tauri/src/events.rs
 */

/**
 * Event names as constants to prevent typos
 */
export const EventNames = {
  // === Chat & AI Workflow ===
  
  /** AI chat message chunk received from zcoderd */
  CHAT_UPDATE: 'chat-update',
  
  /** AI chat response completed */
  CHAT_DONE: 'chat-done',

  /** Hook-driven tool approval request */
  APPROVAL_REQUEST: 'approval-request',
  
  /** Error occurred during chat */
  CHAT_ERROR: 'chat-error',
  
  /** Tool execution requires user confirmation */
  REQUEST_CONFIRMATION: 'request-confirmation',
  
  /** Tool execution started */
  TOOL_EXECUTION_STARTED: 'tool-execution-started',
  
  /** Tool execution completed successfully */
  TOOL_EXECUTION_COMPLETED: 'tool-execution-completed',
  
  /** AI model changed */
  MODEL_CHANGED: 'model-changed',
  
  /** Command execution completed */
  COMMAND_EXECUTED: 'command-executed',
  
  // === File Edit Workflow ===
  
  /** File edit proposed by AI, needs user review */
  PROPOSE_EDIT: 'propose-edit',
  
  /** Change successfully applied to disk */
  CHANGE_APPLIED: 'change-applied',
  
  /** Change rejected by user */
  CHANGE_REJECTED: 'change-rejected',
  
  /** Edit application failed */
  EDIT_FAILED: 'edit-failed',
  
  /** All edits applied successfully (Accept All completed) */
  ALL_EDITS_APPLIED: 'all-edits-applied',
  
  // === File Operations ===
  
  /** File opened in editor */
  FILE_OPENED: 'file-opened',
  
  /** File closed in editor */
  FILE_CLOSED: 'file-closed',
  
  /** File saved to disk */
  FILE_SAVED: 'file-saved',
  
  /** File content modified (unsaved) */
  FILE_MODIFIED: 'file-modified',
  
  /** Active file/tab changed */
  ACTIVE_FILE_CHANGED: 'active-file-changed',
  
  // === Workspace ===
  
  /** Workspace folder changed */
  WORKSPACE_CHANGED: 'workspace-changed',
  
  /** Project files changed (added/deleted) */
  PROJECT_FILES_CHANGED: 'project-files-changed',

  /** Request explorer refresh */
  REFRESH_EXPLORER: 'refresh-explorer',

  // === Terminal ===

  /** Terminal reported a cwd change */
  TERMINAL_CWD_CHANGED: 'terminal-cwd-changed',

  // === Connection & Status ===
  
  /** Connection status to zcoderd changed */
  CONNECTION_STATUS: 'connection-status',
  
  /** General backend error */
  BACKEND_ERROR: 'backend-error',
  
  // === Documents ===
  
  /** Open ephemeral document (research results, etc) */
  OPEN_EPHEMERAL_DOCUMENT: 'open-ephemeral-document',
  
  /** Todo list updated by AI for task progress tracking */
  TODO_UPDATED: 'todo_updated',
} as const;

export interface ChatDonePayload {
  finish_reason: string;
}

/**
 * Payload for chat-error event
 */
export interface ChatErrorPayload {
  code: string;
  error?: string | null;
  message?: string | null;
  detail?: string | null;
  i18nKey?: string | null;
  i18nParams?: Record<string, string> | null;
}

export interface ContextLengthExceededPayload {
  message: string;
  token_count: number | null;
  max_tokens: number | null;
  excess: number | null;
  recoverable: boolean;
  recovery_hint: string | null;
  titleKey?: string | null;
  recoverableHintKey?: string | null;
  nonRecoverableHintKey?: string | null;
}

export interface MessageTooLargePayload {
  message: string;
  recovery_hint: string;
  titleKey?: string | null;
  recoveryHintLabelKey?: string | null;
}

/**
 * Payload for request-confirmation event
 */
export interface RequestConfirmationPayload {
  actions: StructuredAction[];
}

export interface ApprovalRequestPayload {
  session_id: string;
  approval_id: string;
  tool_call_id: string;
  tool_name: string;
  arguments: unknown;
  source?: string | null;
  rule_name?: string | null;
  message?: string | null;
  decision?: string | null;
}

export interface StructuredAction {
  id: string;
  command: string;
  description: string;
  actionKind?: 'command' | 'generic_tool';
  descriptionKey?: string | null;
  descriptionParams?: Record<string, string> | null;
  cwd?: string;
  root_command?: string;
  cwd_outside_workspace?: boolean;
  is_generic_tool: boolean;
}

/**
 * Payload for propose-edit event
 */
export interface ProposeEditPayload {
  id: string;
  path: string;
  old_content: string;
  new_content: string;
}

/**
 * Todo item for task progress tracking
 */
export interface TodoItem {
  content: string;      // Imperative form: "Fix authentication bug"
  activeForm: string;   // Present continuous: "Fixing authentication bug"
  status: 'pending' | 'in_progress' | 'completed';
}

/**
 * Payload for tool-execution-completed event
 */
export interface ToolExecutionCompletedPayload {
  tool_name: string;
  tool_call_id: string;
  success: boolean;
  skipped?: boolean;
}

File: src-tauri/src/events.rs
/// Internal Tauri Events for zblade
///
/// This module defines the event contract between zblade's Rust backend and React frontend.
/// Events flow in one direction: Backend → Frontend (via Tauri's emit)
/// Frontend → Backend communication uses invoke() commands instead.
use serde::{Deserialize, Serialize};
use std::collections::BTreeMap;

/// Event names as constants to prevent typos
pub mod event_names {
    // === Chat & AI Workflow ===

    /// AI chat response completed
    pub const CHAT_DONE: &str = "chat-done";

    /// Hook-driven tool approval request from server
    pub const APPROVAL_REQUEST: &str = "approval-request";

    /// Error occurred during chat
    pub const CHAT_ERROR: &str = "chat-error";

    /// Tool execution requires user confirmation
    pub const REQUEST_CONFIRMATION: &str = "request-confirmation";

    /// Tool execution completed successfully
    pub const TOOL_EXECUTION_COMPLETED: &str = "tool-execution-completed";

    /// Command execution completed
    pub const COMMAND_EXECUTED: &str = "command-executed";

    /// Command execution started (with terminal)
    pub const COMMAND_EXECUTION_STARTED: &str = "command-execution-started";

    // === File Edit Workflow ===

    /// File edit proposed by AI, needs user review
    pub const PROPOSE_EDIT: &str = "propose-edit";

    /// Change successfully applied to disk
    pub const CHANGE_APPLIED: &str = "change-applied";

    // === Workspace ===

    /// Request explorer refresh
    pub const REFRESH_EXPLORER: &str = "refresh-explorer";

    // === Chat Session ===

    /// The chat session id was minted or restored (new/loaded/reset
    /// conversation). Warmup context prefetch re-warms under the new key.
    pub const SESSION_ID_CHANGED: &str = "session-id-changed";

    // === Documents ===

    /// Open ephemeral document (research results, etc)
    pub const OPEN_EPHEMERAL_DOCUMENT: &str = "open-ephemeral-document";

    // === Task Management ===

    /// Todo list updated by AI for task progress tracking
    pub const TODO_UPDATED: &str = "todo_updated";

    // === Terminal ===

    /// Terminal reported a cwd change
    pub const TERMINAL_CWD_CHANGED: &str = "terminal-cwd-changed";

    // === History ===

    /// History entry added (snapshot created)
    pub const HISTORY_ENTRY_ADDED: &str = "history-entry-added";
}

/// Payload for history-entry-added event
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HistoryEntryAddedPayload {
    pub entry: crate::history::HistoryEntry,
}

/// Payload for chat-done event
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatDonePayload {
    pub finish_reason: String,
}

/// Payload for chat-error event
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChatErrorPayload {
    pub code: String,
    pub error: Option<String>,
    pub message: Option<String>,
    pub detail: Option<String>,
    #[serde(rename = "i18nKey")]
    pub i18n_key: Option<String>,
    #[serde(rename = "i18nParams")]
    pub i18n_params: Option<BTreeMap<String, String>>,
}

/// Payload for context-length-exceeded event
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ContextLengthExceededPayload {
    pub message: String,
    pub token_count: Option<u64>,
    pub max_tokens: Option<u64>,
    pub excess: Option<u64>,
    pub recoverable: bool,
    pub recovery_hint: Option<String>,
    #[serde(rename = "titleKey")]
    pub title_key: Option<String>,
    #[serde(rename = "recoverableHintKey")]
    pub recoverable_hint_key: Option<String>,
    #[serde(rename = "nonRecoverableHintKey")]
    pub non_recoverable_hint_key: Option<String>,
}

/// Payload for message-too-large event
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MessageTooLargePayload {
    pub message: String,
    pub recovery_hint: String,
    #[serde(rename = "titleKey")]
    pub title_key: Option<String>,
    #[serde(rename = "recoveryHintLabelKey")]
    pub recovery_hint_label_key: Option<String>,
}

/// Payload for request-confirmation event
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RequestConfirmationPayload {
    pub actions: Vec<StructuredAction>,
}

/// Payload for approval-request event
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApprovalRequestPayload {
    pub session_id: String,
    pub approval_id: String,
    pub tool_call_id: String,
    pub tool_name: String,
    pub arguments: serde_json::Value,
    pub source: Option<String>,
    pub rule_name: Option<String>,
    pub message: Option<String>,
    pub decision: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum StructuredActionKind {
    Command,
    GenericTool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct StructuredAction {
    pub id: String,
    pub command: String,
    pub description: String,
    #[serde(rename = "actionKind")]
    pub action_kind: StructuredActionKind,
    #[serde(rename = "descriptionKey")]
    pub description_key: Option<String>,
    #[serde(rename = "descriptionParams")]
    pub description_params: Option<BTreeMap<String, String>>,
    pub cwd: Option<String>,
    pub root_command: Option<String>,
    pub cwd_outside_workspace: Option<bool>,
    pub is_generic_tool: bool,
}

/// Payload for propose-edit event
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProposeEditPayload {
    pub id: String,
    pub path: String,
    pub old_content: String,
    pub new_content: String,
}

/// Todo item for task progress tracking
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TodoItem {
    pub content: String,
    #[serde(rename = "activeForm")]
    pub active_form: String,
    pub status: String, // 'pending' | 'in_progress' | 'completed' | 'cancelled'
}

/// Payload for todo_updated event
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TodoUpdatedPayload {
    pub todos: Vec<TodoItem>,
}

/// Payload for terminal-cwd-changed event
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TerminalCwdChangedPayload {
    pub id: String,
    pub cwd: String,
}

/// Payload for tool-execution-completed event
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ToolExecutionCompletedPayload {
    pub tool_name: String,
    pub tool_call_id: String,
    pub success: bool,
    #[serde(default)]
    pub skipped: bool,
}

/// Payload for command-executed event
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommandExecutedPayload {
    pub command: String,
    pub cwd: Option<String>,
    pub output: String,
    #[serde(rename = "exitCode")]
    pub exit_code: i32,
    pub duration: Option<u64>,
    pub call_id: String,
}

/// Payload for change-applied event
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ChangeAppliedPayload {
    pub change_id: String,
    pub file_path: String,
    #[serde(default)]
    pub file_paths: Vec<String>,
}

/// Payload for command-execution-started event
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CommandExecutionStartedPayload {
    pub command_id: String,
    pub call_id: String,
    pub command: String,
    pub program: Option<String>,
    pub args: Option<Vec<String>>,
    pub shell: Option<bool>,
    pub cwd: Option<String>,
    pub blocking: bool,
    pub wait_ms_before_async: Option<u64>,
}
